File: 프로그래머스/3/81303. 표 편집/표 편집.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

string solution(int n, int k, vector<string> cmd) {
    string answer(n, 'O');
    
    vector<pair<int, int>> adj;
    for(int i = 0; i < n; i++) {
        adj.push_back({i - 1, i + 1});
    }
    stack<int> deleted;
    
    for(string c: cmd) {
        if(c[0] == 'U') {
            int dist = stoi(c.substr(2));
            while(dist--) {
                k = adj[k].first;
            }
        }
        else if(c[0] == 'D') {
            int dist = stoi(c.substr(2));
            while(dist--) {
                k = adj[k].second;
            }            
        }
        else if(c[0] == 'C') {
            answer[k] = 'X';
            deleted.push(k);
            int prevIdx = adj[k].first;
            int nxtIdx = adj[k].second;

            if(prevIdx != -1) adj[prevIdx].second = nxtIdx;
            if(nxtIdx != n) adj[nxtIdx].first = prevIdx;

            if(nxtIdx == n) k = prevIdx;
            else k = nxtIdx;    
        }
        else if(c[0] == 'Z') {
            if(!deleted.empty()) {
                int idx = deleted.top();
                deleted.pop();
                answer[idx] = 'O';

                int prevIdx = adj[idx].first;
                int nxtIdx = adj[idx].second;

                if(prevIdx != -1) adj[prevIdx].second = idx;
                if(nxtIdx != n) adj[nxtIdx].first = idx;
            }
        }
    }    
    
    return answer;
}

File: 프로그래머스/3/81303. 표 편집/표 편집.py
from collections import deque

def solution(n, k, cmd):
    answer = ['O'] * n
    
    adjs = [[idx - 1, idx + 1] for idx in range(n)]
    deleted = deque()
    
    for cur_cmd in cmd:
        if cur_cmd[0] == 'U':
            x = int(cur_cmd[2:])
            for _ in range(x):
                k = adjs[k][0]
                
        elif cur_cmd[0] == 'D':
            x = int(cur_cmd[2:])
            for _ in range(x):
                k = adjs[k][1]
                
        elif cur_cmd[0] == 'C':
            answer[k] = 'X'
            deleted.append(k)
            
            up, down = adjs[k]
            
            if up != -1:
                adjs[up][1] = down
                
            if down != n:
                adjs[down][0] = up
                
            if down == n:
                k = up
            else:
                k = down
            
        elif cur_cmd[0] == 'Z':
            re = deleted.pop()
            answer[re] = 'O'
            
            re_up, re_down = adjs[re]
            
            if re_up != -1:
                adjs[re_up][1] = re
            
            if re_down != n:
                adjs[re_down][0] = re
    
    answer = "".join(answer)
    return answer